### PR TITLE
Add cert monitoring

### LINKED ISSUE #65: Add monitoring of certificates about to expire
within certain time periods.

### PR DESCRIPTION
**Pull request by woutslakhorst**
Closes #65

### CODE REVIEW

File: pkg/crypto.go [hunk 1]
```diff
@@ -145,6 +146,49 @@ func (k opaquePrivateKey) Sign(rand io.Reader, digest []byte, opts crypto.Signer
...
+			"4_weeks",
+		},
+	}
+
+	for _, p := range periods {
+		m := storage.CertificateMonitor{
+			Period:      p.period,
+			PeriodLabel: p.label,
+			Storage:     client.Storage,
+		}
```
Comment (reinkrul) [inline pkg/crypto.go]: isn't this something to move to the `storage` package as `DefaultCertificateMonitors` or something?

Or at least an initialization function in `storage` (`NewCertificateMonitor(...)`)

File: pkg/storage/metrics.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+// in minutes
+const checkInterval = 15
+
+const label = "period"
+
+// ErrAlreadyStarted is returned when start is called on an already running monitor
+var ErrAlreadyStarted = errors.New("certificate monitor already started")
+
+// we register only one
+var metric *prometheus.GaugeVec
```
Comment (reinkrul) [inline pkg/storage/metrics.go:39]: ugh more hidden, shared state making unit testing hard(er). Can't we solve this by making Register() idempotent (by making it only register it when it's not already registered)

File: pkg/storage/metrics.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+type CertificateMonitor struct {
+	exit        chan bool
+	Period      time.Duration
+	PeriodLabel string
+	Storage     Storage
+}
+
+// Start the certificate monitor for checking expiring certificates between now and the configure period.
+func (cm *CertificateMonitor) Start() error {
+
```
Comment (reinkrul) [inline pkg/storage/metrics.go]: random newlines?

File: pkg/storage/metrics.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	"sync"
+	"time"
+
+	"github.com/nuts-foundation/nuts-crypto/log"
+	errors2 "github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// in minutes
+const checkInterval = 15
```
Comment (reinkrul) [inline pkg/storage/metrics.go:31]: you could use `time.Duration` here (though not sure if you can still make it a const, then)

File: pkg/storage/metrics.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+			cm.checkExpiry()
+		}
+	}
+}
+
+func (cm CertificateMonitor) checkExpiry() {
+	expiringCerts, err := cm.Storage.GetExpiringCertificates(time.Now(), time.Now().Add(cm.Period))
+
+	if err != nil {
+		log.Logger().Error(errors2.Wrap(err, "failed to check expiry"))
```
Comment (reinkrul) [inline pkg/storage/metrics.go]: why not just:

`log.Logger().Error("Failed to check expiry:", err)`

File: pkg/storage/metrics.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+var ErrAlreadyStarted = errors.New("certificate monitor already started")
+
+// we register only one
+var metric *prometheus.GaugeVec
+var mutex sync.Mutex
+
+// CertificateMonitor represents a go procedure which monitors expiring certificates within a given period.
+type CertificateMonitor struct {
+	exit        chan bool
+	Period      time.Duration
```
Comment (reinkrul) [inline pkg/storage/metrics.go]: if you initialize this struct from this package (as suggested earlier) you can make these fields private, making it more or less immutable.

File: pkg/storage/metrics.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+		)
+		err = prometheus.Register(metric)
+	}
+
+	return err
+}
+
+// Stop the certificate monitor
+func (cm *CertificateMonitor) Stop() {
+	cm.exit <- true
```
Comment (reinkrul) [inline pkg/storage/metrics.go:120]: I'm actually not entirely sure whether this adds something. Blocking goroutines (waiting for tickers to fire) don't block the process from exiting and since we don't wait here for `loop()` to end graceful shutdown becomes subject for racy conditions (if the process exits before the `exit` channel has been read and shut down, I'm not considering it a 'graceful' shutdown, even when that shutdown doesn't actually do anything other than returning from the `loop()` function).

We have this construct all over the place though...

**Review by reinkrul — CHANGES_REQUESTED** [review]

**Comment by woutslakhorst** [comment]
- [x] Add monitoring docs

**Comment by woutslakhorst** [comment]
Screenshot 2020-07-14 at 08 47 55

File: pkg/storage/metrics.go (hunk 7 above)
Comment (woutslakhorst) [reply pkg/storage/metrics.go:120]: without it, it blocks (at least the tests)

File: pkg/storage/metrics.go (hunk 4 above)
Comment (woutslakhorst) [reply pkg/storage/metrics.go:31]: then it would become a var

File: pkg/crypto.go (hunk 1 above)
Comment (woutslakhorst) [reply pkg/crypto.go]: moved it to DefaultCertificateMonitors and made attrs private

File: pkg/storage/metrics.go (hunk 2 above)
Comment (woutslakhorst) [reply pkg/storage/metrics.go:39]: Unfortunately you can't get a reference to a collector via the Prometheus Registry, you'll get an interface type. You could do some casting, but that's rather ugly as well. So you'll have to store your own reference to the collector that has been registered.

Another challenge is that collectors are compared based on the name, not the supported labels. Only allowing the first to register would yield unexpected behaviour. Specifying which metrics are registered  is done at compile-time. So returning an error when this is done twice seems appropriate.

File: pkg/storage/metrics.go (hunk 6 above)
Comment (woutslakhorst) [reply pkg/storage/metrics.go]: done

File: engine/engine.go [hunk 8]
```diff
@@ -63,7 +66,7 @@ func flagSet() *pflag.FlagSet {
 	flags.String(types.ConfigMode, defs.Mode, fmt.Sprintf("Server or client, when client it uses the HttpClient, default: %s", defs.Mode))
 	flags.String(types.ConfigAddress, defs.Address, fmt.Sprintf("Interface and port for http server to bind to, default: %s", defs.Address))
 	flags.Int(types.ConfigClientTimeout, defs.ClientTimeout, fmt.Sprintf("Time-out for the client in seconds (e.g. when using the CLI), default: %d", defs.ClientTimeout))
-	flags.String(types.ConfigStorage, defs.Storage, fmt.Sprintf("Storage to use, 'fs' for file system, default: %s", defs.Storage))
+	flags.String(types.ConfigStorage, defs.Storage, fmt.Sprintf("storage to use, 'fs' for file system, default: %s", defs.Storage))
```
Comment (reinkrul) [inline engine/engine.go]: why this change?

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: engine/engine.go (hunk 8 above)
Comment (woutslakhorst) [reply engine/engine.go]: reverted

**Review by reinkrul — APPROVED** [review]
